src/Workspaces/Core/Portable/Storage/SQLite/Interop/SafeSqliteBlobHandle.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using SQLitePCL;
+
+namespace Microsoft.CodeAnalysis.SQLite.Interop
+{
+    internal sealed class SafeSqliteBlobHandle : SafeHandle
+    {
+        private readonly sqlite3_blob? _wrapper;
+        private readonly SafeHandleLease _lease;
+
+        public SafeSqliteBlobHandle(SafeSqliteHandle sqliteHandle, sqlite3_blob? wrapper)
+            : base(invalidHandleValue: IntPtr.Zero, ownsHandle: true)
+        {
+            _wrapper = wrapper;
+            SetHandle(wrapper?.ptr ?? IntPtr.Zero);
```

Comment:
Do you care about thread aborts?  If the answer is no, no problem.  If the answer is yes, and if you were to incur one right here, you'd have a SafeHandle whose finalizer is going to sqlite3_blob_close _wrapper even though you've not taken a lease on it.

Reply:
The `SafeHandle.handle` defaults to `IntPtr.Zero`, which this type has indicated is the invalid handle value. Would that not cause the finalizer to skip calling `ReleaseHandle`?

Reply:
I meant after the call to SetHandle and before the setting of _lease.  At that point your handle is valid but you've not taken out the lease.

Reply:
I believe this is the correct order. If it were the other way around, the lease would never be returned and the underlying connection never closed.

Reply:
It may very well be.  I'm simply calling out that there's a potential issue with this ordering if you care about thread aborts.  It's fine to say you don't :)

Reply:
> I believe this is the correct order. If it were the other way around, the lease would never be returned

In a thread abort during this ctor you still will never return the lease, because IsInvalid is true, so ReleaseHandle won't get called.  You'd need to do a CER if you really care about thread abort... and if not, either order is fine.

Reply:
> ReleaseHandle won't get called

ReleaseHandle will still get called by the finalizer right?

Reply:
* `Dispose()` => `Dispose(true); GC.SuppressFinalize(this);`
* `Finalize()` => `Dispose(false)`
* `Dispose(bool)` => `if (0 == Interlocked.Exchange(ref _disposed, 1) { DangerousRelease(); }`
* `DangerousRelease()` => `if (Interlocked.Decrement(ref _count) == 0 && !IsInvalid && ownsHandle) { ReleaseHandle(); }`


Modulo details.

So the current order would be:

* SafeSqlliteBlobHandle..ctor(...)
  * SafeHandle..ctor(IntPtr.Zero, true)
    * handle = IntPtr.Zero;
  * _wrapper = wrapper
  * external Thread.Abort();
* ~SafeHandle
  * Dispose(false)
    * DangerousRelease
      * IsInvalid returns true.  Moving right along...

If you passed the handle in the chained ctor, or otherwise set handle first

* SafeSqlliteBlobHandle..ctor(...)
  * SafeHandle..ctor(IntPtr.Zero, true)
    * handle = IntPtr.Zero;
  * SetHandle(wrapper?.ptr ?? IntPtr.Zero)
    * handle = x;
  * external Thread.Abort();
* ~SafeHandle
  * Dispose(false)
    * DangerousRelease
      * IsInvalid returns false
      * ReleaseHandle
        * `wrapper` is null in ReleaseHandle, what sort of problems is that going to cause?

One leaks a resource, one possibly results in ReleaseHandle returning false, which causes an exception, which kills the finalizer thread and the whole process.  Pick your poison.  Or just celebrate in the fact that .NET Core got rid of thread aborts.